railgun/internal/proxy/clientgo_cached_proxy_resolver.go

```diff
@@ -216,18 +217,25 @@ func (p *clientgoCachedProxyResolver) startCacheServer() {
 				p.logger.Error(err, "object could not be updated in the cache and will be discarded")
 				break
 			}
+			backendNeedsSync = true
 		case cobj := <-p.del:
 			if err := p.cacheDelete(cobj); err != nil {
 				p.logger.Error(err, "object could not be deleted from the cache and will be discarded")
 				break
 			}
+			backendNeedsSync = true
 		case <-p.syncTicker.C:
+			if !p.enableReverseSync && !backendNeedsSync {
+				break
+			}
+
 			updateConfigSHA, err := p.kongUpdater(p.ctx, p.lastConfigSHA, p.cache, p.ingressClassName, p.deprecatedLogger, p.kongConfig, p.enableReverseSync)
 			if err != nil {
 				p.logger.Error(err, "could not update kong admin")
 				break
 			}
 			p.lastConfigSHA = updateConfigSHA
+			backendNeedsSync = false
```

Comment:
Theoretically, this could reduce proxy sync frequency possibly. 

The real problem is kongUpdater does not get back within a cycle due to the large data within one POST /config, and reconcile keeps pinging OnUpdate, the frequency is still high.

Reply:
> Theoretically, this could reduce proxy sync frequency possibly.

This statement is in contradiction of my understanding and the intentions of this change:

With this change:

1. any successful update to the cache will flag proxy sync as needed
2. syncs will only ever be disabled if the `kongUpdater` synchronizes the current cache to the proxy, _without error_

In this scenario, the `p.syncTicker` will trigger an proxy sync _any time the cache changes_ and then flag sync off only once that cache is resolved to the proxy.

This all happens synchronously within the cacheServer goroutine meaning that there are _no separate routines competing for the configuration of the `backendNeedsSync` variable_.

This does _reduce the number of calls to kongUpdater()_ but that is _intentional_ because we don't need to call it if the following conditions are both true:

a. no changes to the cache have been seen
b. the last `kongUpdater()` call to synchronize data to the proxy was successful.

> The real problem is kongUpdater does not get back within a cycle due to the large data within one POST /config, and reconcile keeps pinging OnUpdate, the frequency is still high.

If there's a large amount of data in a single `POST /config` and the time it takes to complete the `POST` takes longer than the interval for the `p.syncTicker` (which is a `time.Ticker`) then the `cacheServer` will _not_ trigger any more `kongUpdater` calls until the current one finishes, as all of this is done _synchronously_ within the goroutine.

Here is example code that should help to illustrate my meaning:

https://play.golang.org/p/Vvw8qVeHNnq

In the example code, the result will be:

```
length of ticker channel 1
count is 1
```

A call will not be made every 1 second, as this is not how Go channels or the `time.Ticker` implementation works.

In the implementation of the `time.Ticker`, and in our usage of it if we have long running calls to `kongUpdater` that are taking longer than the tick time, the ticket will just "be on hold", and in fact will drop tickets (by design):

https://cs.opensource.google/go/go/+/refs/tags/go1.16.6:src/time/tick.go;l=26

```
	// Give the channel a 1-element time buffer.
	// If the client falls behind while reading, we drop ticks
	// on the floor until the client catches up.
```

---

Let me know if that is helpful and illustrative, or if you think I've overlooked or misunderstood something and you still see a problem here please help me understand the steps to produce the problem and then lets get it fixed.